ext/fg/js/popup.js

```diff
@@ -207,20 +207,17 @@ class Popup {
             const parentFrameId = (typeof this._frameId === 'number' ? this._frameId : null);
             this._container.setAttribute('src', chrome.runtime.getURL('/fg/float.html'));
             this._container.addEventListener('load', () => {
-                const uniqueId = yomichan.generateId(32);
-                Popup._listenForDisplayPrepareCompleted(uniqueId, resolve);
+                this._listenForDisplayPrepareCompleted(resolve);
```

Comment:
There is a timing issue now that uniqueId has been removed and only the popup `id` is used. `id` is unique within a frame, but not unique within a tab, and the `'popupPrepareCompleted'` message is sent to all frames on a tab. `uniqueId` was used to provide uniqueness within a tab.

If two popups are opened at the same time in different frames on the same tab, the body of `_listenForDisplayPrepareCompleted.runtimeMessageCallback` can be triggered for the wrong frame's popup. In practice this is extremely rare to the point of being improbable, and the negative effects are even less likely to show up due to how quickly the messages are processed, but the issue is technically still there.

The following example demonstrates that IDs are not unique within a tab. Note that `prepared: 2` is printed, which indicates the timing issue _could_ occur.

```js
_listenForDisplayPrepareCompleted(resolve) {
    const runtimeMessageCallback = ({action, params}, sender, callback) => {
        setTimeout(() => { // added
        if (
            action === 'popupPrepareCompleted' &&
            isObject(params) &&
            params.targetPopupId === this._id
        ) {
            this._prepareCount = (this._prepareCount || 0) + 1; // added
            console.log('prepared: ' + this._prepareCount); // added
            chrome.runtime.onMessage.removeListener(runtimeMessageCallback);
            callback();
            resolve();
            return false;
        }
        }, 2000); // added
    };
    chrome.runtime.onMessage.addListener(runtimeMessageCallback);
}
/*
Prints:
prepared: 1
prepared: 2
prepared: 1
*/
```

Compared with the old method using `uniqueId`:

```js
static _listenForDisplayPrepareCompleted(uniqueId, resolve) {
    const runtimeMessageCallback = ({action, params}, sender, callback) => {
        setTimeout(() => { // added
        if (
            action === 'popupPrepareCompleted' &&
            typeof params === 'object' &&
            params !== null &&
            params.uniqueId === uniqueId
        ) {
            this._prepareCount = (this._prepareCount || 0) + 1; // added
            console.log('prepared: ' + this._prepareCount); // added
            chrome.runtime.onMessage.removeListener(runtimeMessageCallback);
            callback();
            resolve();
            return false;
        }
        }, 2000); // added
    };
    chrome.runtime.onMessage.addListener(runtimeMessageCallback);
}
/*
Prints:
prepared: 1
prepared: 1
*/
```

Reply:
Thanks for noticing this, I was under the impression that iframe popup IDs are also unique within the root page popup proxy host because the last time I consciously scanned iframes was when doing https://github.com/FooSoft/yomichan/pull/364. It actually doesn't even assign an ID for the iframe popups and guesses the position instead.

The simplest solution would be to generate an unique ID for each popup and use it instead of the sequence, but I'd like to centralize tab communication at some point (not this PR) so that the root page popup proxy is aware of popups within iframes too.

Reply:
I read the code a bit, and it could be possible to drop `messageToken` as well. I remember I suggested it to come from `Backend`, but currently it doesn't seem to be used in a way that would require it to be the same among different tabs.

Reply:
...except in the case of `prepare`. I'll leave it there :smile: